synapse/logging/_structured.py

```diff
@@ -0,0 +1,222 @@
+    ILogObserver,
+    Logger,
+    LogLevel,
+    LogLevelFilterPredicate,
+    LogPublisher,
+    eventAsText,
+    globalLogBeginner,
+    jsonFileLogObserver,
+)
+
+from synapse.config._base import ConfigError
+from synapse.logging.context import LoggingContext
+
+
+def stdlib_log_level_to_twisted(level):
+    """
+    Convert a stdlib log level to Twisted's log level.
+    """
+    lvl = level.lower().replace("warning", "warn")
+    return LogLevel.levelWithName(lvl)
+
+
+@attr.s
+@implementer(ILogObserver)
+class LogContextObserver(object):
+    observer = attr.ib()
+
+    def __call__(self, event):
+
+        # Filter out some useless events that Twisted outputs
+        if "log_text" in event:
+            if event["log_text"].startswith("DNSDatagramProtocol starting on "):
+                return
+
+            if event["log_text"].startswith("(UDP Port "):
+                return
+
+            if event["log_text"].startswith("Timing out client"):
+                return
+
+        event["request"] = "None"
+        event["scope"] = None
+
+        context = LoggingContext.current_context()
+        if context is not None and isinstance(context, LoggingContext):
+            event["request"] = context.request
+            event["scope"] = context.scope
+
+        return self.observer(event)
+
+
+class PythonStdlibToTwistedLogger(logging.Handler):
+    """
+    Transform a Python stdlib log message into a Twisted one.
+    """
+
+    def __init__(self, observer, *args, **kwargs):
+        self.observer = observer
+        super().__init__(*args, **kwargs)
+
+    def emit(self, record):
+
+        self.observer(
+            {
+                "log_time": record.created,
+                "log_text": record.getMessage(),
+                "log_format": "{log_text}",
+                "log_namespace": record.name,
+                "log_level": stdlib_log_level_to_twisted(record.levelname),
+            }
+        )
+
+
+def SynapseFileLogObserver(outFile):
+    """
+    A log observer that formats events like Synapse wants and sends them to
+    `outFile`.
+    """
+
+    def formatEvent(_event):
+        event = _event.copy()
+        event["log_level"] = event["log_level"].name.upper()
+        event["log_format"] = (
+            "- {log_namespace} - {log_level} - {request} - "
+            + event.get("log_format", "{log_text}")
+        )
+        return eventAsText(event, includeSystem=False) + "\n"
+
+    return FileLogObserver(outFile, formatEvent)
+
+
+class LoggingOutputType(Names):
+    CONSOLE = NamedConstant()
+    JSON = NamedConstant()
+    FLUENTD = NamedConstant()
+
+
+class OutputPipeType(Values):
+    stdout = ValueConstant(sys.__stdout__)
+    stderr = ValueConstant(sys.__stderr__)
+
+
+@attr.s
+class HandlerConfiguration(object):
+    name = attr.ib()
+    type = attr.ib()
+    location = attr.ib()
+
+
+def parse_handler_configs(config):
+    """
+    Parse the logging format version 2 handler section.
```

Comment:
can you document the args and return types please?

Reply:
done